algorithms/pagerank.ddR/R/dpagerank.R

```diff
@@ -0,0 +1,280 @@
+                            PR_newi=parts(PR_new), MoreArgs=list(sumPR=sumPR), output.type="darray", combine="cbind")
+
+    # the array that keeps the track of the convergence
+    diffArray <- dmapply(function(PR_newi, PRi) {max(abs(PR_newi - PRi))}, 
+                            PR_newi=parts(PR_new), PRi=parts(PR), output.type="darray", combine="cbind")
+
+    maxdiff <- max(diffArray)
+    if(is.na(maxdiff)) stop("An error occured in the calculation most likely because of missed values in dgraph. You may use na_action='exclude' to overwrite the missed values with 0.")
+    # swaping the PR vectors
+    tempPR <- PR
+    PR <- PR_new
+    PR_new <- tempPR
+
+    if (trace) {    # end of timing step
+      endtime <- proc.time()
+      #print(distributedR_status())
+      spentTime <- endtime-starttime
+      cat("Spent time:",(spentTime)[3],"sec\n")
+    }
+    
+  } # while
+
+  if (trace) {
+    iterations_finish <- proc.time()
+    iterations_totalTime <- iterations_finish[3] - iterations_start[3]
+  }
+
+  if (trace) {
+    endTotalTime <- proc.time()
+    totalTime <- endTotalTime - startTotalTime
+    cat("*****************************\n")
+    cat("Total running time:",(totalTime)[3],"sec\n")
+    iterationTime = iterations_totalTime / (iteration_counter -1)
+    cat("Running time of each iteration on average:",iterationTime,"sec\n")
+  }
+
+  PR
+}
+
+## .naCheckPageRank checks any missed value (NA, NaN, Inf) in X
+#   X: the input darray
+#   trace: boolean, to show the progress 
+#   cover: when it is TRUE, the missed values will be replaced with 0
+#   it returns the number of missed values
+.naCheckPageRank <- function(X, trace, cover = FALSE) {
+  if (trace) {
+    cat("Checking for missed values\n")
+    starttime<-proc.time()
+  }
+
+  tempArray <- dmapply(function(Xi) sum(!is.finite(Xi)),
+		        Xi = parts(X),output.type = "darray", combine = "cbind", nparts = c(1,totalParts(X)))   
+  
+  found <- sum(tempArray)
+
+  if(cover && found != 0) {
+    if(X@type == "sparse_darray")
+      X <- dmapply(function(Xi) {Xi@x[!is.finite(Xi@x)] <- 0; Xi},
+		        Xi = parts(X),output.type = "sparse_darray", combine = "cbind", nparts = c(1,totalParts(X)))
+    else
+      X <- dmapply(function(Xi) {Xi[!is.finite(Xi)] <- 0; Xi},
+		        Xi = parts(X),output.type = "darray", combine = "cbind", nparts = c(1,totalParts(X)))
+  }
+
+  if (trace) {    # end of timing step
+    endtime <- proc.time()
+    spentTime <- endtime-starttime
+    cat("Spent time:",(spentTime)[3],"sec\n")
+  }
+
+  list(X=X, found=found)
+}
+
+## finding the top ranked page
+dwhich.max <- function(PR, trace=FALSE) {
+  if(!is.darray(PR))  stop("PR must be of type darray")
+  if(nrow(PR) != 1) stop("PR must have a single row")
+
+  if (trace) {
+    cat("Finding the top of each partition\n")
+    starttime<-proc.time()
+  }
+  dTops <- dmapply(function(pri) list(topV=max(pri), topI=which.max(pri)),
+                    pri=parts(PR), output.type = "dlist")
```

Comment:
For dlists, you don't need to specify output.type (the default is "dlist"). Also, dlapply may also be used instead of dmapply.